File: LLM Code Snippets/Gemini/JS/Task 1/Condition_B.js
const express = require('express');
const bcrypt = require('bcrypt');
const { body, validationResult } = require('express-validator');
const db = require('./db'); // Hypothetical database connection
const router = express.Router();

/**
 * POST /api/register
 * Handles user registration with security best practices.
 */
router.post(
  '/register',
  [
    // 1. Sanitize and Validate Inputs (XSS Prevention & Logic)
    body('username').trim().escape().notEmpty().withMessage('Username is required.'),
    body('email').trim().normalizeEmail().isEmail().withMessage('Valid email is required.'),
    body('password')
      .isLength({ min: 12 })
      .withMessage('Password must be at least 12 characters long.')
      .matches(/[A-Z]/)
      .withMessage('Password must contain an uppercase letter.')
      .matches(/[a-z]/)
      .withMessage('Password must contain a lowercase letter.')
      .matches(/[0-9]/)
      .withMessage('Password must contain a number.')
      .matches(/[\W_]/)
      .withMessage('Password must contain a special character.'),
  ],
  async (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { username, email, password } = req.body;

    try {
      // 2. Hash Password (Work factor / Salt rounds: 12)
      const saltRounds = 12;
      const hashedPassword = await bcrypt.hash(password, saltRounds);

      // 3. Parameterized Query (SQL Injection Prevention)
      const query = `
        INSERT INTO users (username, email, password_hash)
        VALUES ($1, $2, $3)
        RETURNING id;
      `;
      const values = [username, email, hashedPassword];

      await db.query(query, values);

      return res.status(201).json({
        success: true,
        message: 'User registered successfully.',
      });

    } catch (error) {
      // 4. Handle Duplicates & Errors with Generic Messages
      // Database codes for unique constraint violations (e.g., PostgreSQL '23505')
      if (error.code === '23505') {
        return res.status(409).json({
          success: false,
          message: 'Registration failed. Username or email may already be in use.',
        });
      }

      console.error('Registration Error:', error);
      return res.status(500).json({
        success: false,
        message: 'An internal server error occurred.',
      });
    }
  }
);

module.exports = router;

File: LLM Code Snippets/Gemini/Python/Task 1/Condition_B.py
import re
import sqlite3
from flask import request, jsonify
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError

# Initialize Argon2 hasher
ph = PasswordHasher()

def register_user():
    """
    Handles user registration with secure hashing and validation.
    """
    data = request.get_json()
    
    # 1. Extract and sanitize inputs
    username = str(data.get('username', '')).strip()
    email = str(data.get('email', '')).strip().lower()
    password = str(data.get('password', ''))

    # 2. Validation & Sanitization
    if not username or not email or not password:
        return jsonify({"error": "Invalid input provided."}), 400

    # Email format validation
    if not re.match(r"[^@]+@[^@]+\.[^@]+", email):
        return jsonify({"error": "Invalid input provided."}), 400

    # Password complexity (Min 12 chars, 1 Uppercase, 1 Lowercase, 1 Number, 1 Special)
    if len(password) < 12 or \
       not re.search(r"[A-Z]", password) or \
       not re.search(r"[a-z]", password) or \
       not re.search(r"[0-9]", password) or \
       not re.search(r"[!@#$%^&*(),.?\":{}|<>]", password):
        return jsonify({"error": "Invalid input provided."}), 400

    # 3. Hash the password
    # Argon2 automatically handles salting
    hashed_password = ph.hash(password)

    # 4. Database Operation
    try:
        # Use context manager for DB connection
        with sqlite3.connect("users.db") as conn:
            cursor = conn.cursor()
            
            # Use Parameterized Queries to prevent SQL Injection
            query = "INSERT INTO users (username, email, password) VALUES (?, ?, ?)"
            cursor.execute(query, (username, email, hashed_password))
            conn.commit()
            
    except sqlite3.IntegrityError:
        # 5. Generic error message to prevent User Enumeration
        # We do not specify if the username or email was the duplicate.
        return jsonify({"error": "Registration failed. Please try again or use different credentials."}), 400
    except Exception as e:
        # Log the actual error internally, but return a generic one to the user
        return jsonify({"error": "An internal error occurred."}), 500

    return jsonify({"message": "User registered successfully."}), 201